Runtime panic when performing any search in version 3.32.0

I just updated to v3.32.0 and it keeps crashing and restarting any time a try to do a search.

**Sourcegraph version:**
3.32.0

**Platform information:**
Debian Linux
Kernel 5.13.4-x86_64-linode146
Linode
docker-ce 5:20.10.8~3-0~debian-buster
Single Sourcegraph container + lang-typescript container

#### Steps to reproduce:

1. Start Sourcegraph v3.32.0
2. Perform a search or type in the search box

#### Expected behavior:

The search should complete.

#### Actual behavior:

When performing a search I see this stack trace in the logs:
```
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x30 pc=0x1b11221]
goroutine 995 [running]:
github.com/sourcegraph/sourcegraph/internal/search/zoekt.doZoektSearchGlobal({0x66f7e00, 0xc001422d80}, {0x6691c40, 0xc00212d560}, {0x293eaad, 0x4}, {0x0, 0x0}, 0x1f4, {0x812c6f8, ...}, ...)
   github.com/sourcegraph/sourcegraph/internal/search/zoekt/indexed_search.go:485 +0x401
github.com/sourcegraph/sourcegraph/internal/search/zoekt.(*IndexedUniverseSearchRequest).Search(0xc000ee94a0, {0x66f7e00, 0xc001422d80}, {0x6694120, 0xc00220ac80})
   github.com/sourcegraph/sourcegraph/internal/search/zoekt/indexed_search.go:173 +0x177
github.com/sourcegraph/sourcegraph/internal/search/unindexed.SearchFilesInRepos.func2()
   github.com/sourcegraph/sourcegraph/internal/search/unindexed/unindexed.go:72 +0x32
golang.org/x/sync/errgroup.(*Group).Go.func1()
   golang.org/x/sync@v0.0.0-20210220032951-036812b2e83c/errgroup/errgroup.go:57 +0x67
created by golang.org/x/sync/errgroup.(*Group).Go
   golang.org/x/sync@v0.0.0-20210220032951-036812b2e83c/errgroup/errgroup.go:54 +0x92
```

If I type in the search box I get the following error:
```
goroutine panic: runtime error: invalid memory address or nil pointer dereference
goroutine 886 [running]:
runtime/debug.Stack()
   runtime/debug/stack.go:24 +0x65
github.com/sourcegraph/sourcegraph/internal/goroutine.Go.func1.1()
   github.com/sourcegraph/sourcegraph/internal/goroutine/goroutine.go:19 +0x39
panic({0x25398c0, 0x80b5e60})
   runtime/panic.go:1038 +0x215
github.com/sourcegraph/sourcegraph/internal/search/zoekt.doZoektSearchGlobal({0x66f8df8, 0xc002037340}, {0x6691c40, 0xc001d16c90}, {0x295797c, 0x6}, {0x0, 0x0}, 0x1e, {0x812c6f8, ...}, ...)
   github.com/sourcegraph/sourcegraph/internal/search/zoekt/indexed_search.go:485 +0x401
github.com/sourcegraph/sourcegraph/internal/search/zoekt.(*IndexedUniverseSearchRequest).Search(0xc0011943c0, {0x66f8df8, 0xc002037340}, {0x6694120, 0xc002037360})
   github.com/sourcegraph/sourcegraph/internal/search/zoekt/indexed_search.go:173 +0x177
github.com/sourcegraph/sourcegraph/internal/search/symbol.Search.func2()
   github.com/sourcegraph/sourcegraph/internal/search/symbol/symbol.go:71 +0xa7
github.com/sourcegraph/sourcegraph/internal/goroutine.Go.func1()
   github.com/sourcegraph/sourcegraph/internal/goroutine/goroutine.go:24 +0x3f
created by github.com/sourcegraph/sourcegraph/internal/goroutine.Go
   github.com/sourcegraph/sourcegraph/internal/goroutine/goroutine.go:16 +0x5b
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x30 pc=0x1b11221]
goroutine 863 [running]:
github.com/sourcegraph/sourcegraph/internal/search/zoekt.doZoektSearchGlobal({0x66f7e00, 0xc001e1b300}, {0x6691c40, 0xc001c7ab88}, {0x293eaad, 0x4}, {0x0, 0x0}, 0x1e, {0x812c6f8, ...}, ...)
   github.com/sourcegraph/sourcegraph/internal/search/zoekt/indexed_search.go:485 +0x401
github.com/sourcegraph/sourcegraph/internal/search/zoekt.(*IndexedUniverseSearchRequest).Search(0xc000191b30, {0x66f7e00, 0xc001e1b300}, {0x6694120, 0xc002008f20})
   github.com/sourcegraph/sourcegraph/internal/search/zoekt/indexed_search.go:173 +0x177
github.com/sourcegraph/sourcegraph/internal/search/unindexed.SearchFilesInRepos.func2()
   github.com/sourcegraph/sourcegraph/internal/search/unindexed/unindexed.go:72 +0x32
golang.org/x/sync/errgroup.(*Group).Go.func1()
   golang.org/x/sync@v0.0.0-20210220032951-036812b2e83c/errgroup/errgroup.go:57 +0x67
created by golang.org/x/sync/errgroup.(*Group).Go
   golang.org/x/sync@v0.0.0-20210220032951-036812b2e83c/errgroup/errgroup.go:54 +0x92
```

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -23,6 +23,8 @@ All notable changes to Sourcegraph are documented in this file.
 
 ### Fixed
 
+- A search regression in 3.32.0 which caused instances with search indexing _disabled_ (very rare) via `"search.index.enabled": false,` in their site config to crash with a panic. [#25321](https://github.com/sourcegraph/sourcegraph/pull/25321)
+- An issue where the default `search.index.enabled` value on single-container Docker instances would incorrectly be computed as `false` in some situations. [#25321](https://github.com/sourcegraph/sourcegraph/pull/25321)
 - StatefulSet service discovery in Kubernetes correctly constructs pod hostnames in the case where the ServiceName is different from the StatefulSet name. [#25146](https://github.com/sourcegraph/sourcegraph/pull/25146)
 
 ### Removed

--- a/internal/conf/computed.go
+++ b/internal/conf/computed.go
@@ -4,7 +4,6 @@ import (
 	"context"
 	"log"
 	"os"
-	"strconv"
 	"strings"
 
 	"github.com/sourcegraph/sourcegraph/internal/api"
@@ -204,19 +203,12 @@ func UpdateChannel() string {
 }
 
 // SearchIndexEnabled returns true if sourcegraph should index all
-// repositories for text search. If the configuration is unset, it returns
-// false for the docker server image (due to resource usage) but true
-// elsewhere. Additionally it also checks for the outdated environment
-// variable INDEXED_SEARCH.
+// repositories for text search.
 func SearchIndexEnabled() bool {
 	if v := Get().SearchIndexEnabled; v != nil {
 		return *v
 	}
-	if v := os.Getenv("INDEXED_SEARCH"); v != "" {
-		enabled, _ := strconv.ParseBool(v)
-		return enabled
-	}
-	return DeployType() != DeploySingleDocker
+	return true // always on by default in all deployment types, see confdefaults.go
 }
 
 func BatchChangesEnabled() bool {

--- a/internal/conf/computed_test.go
+++ b/internal/conf/computed_test.go
@@ -38,16 +38,6 @@ func TestSearchIndexEnabled(t *testing.T) {
 		sc:   &Unified{SiteConfiguration: schema.SiteConfiguration{SearchIndexEnabled: boolPtr(false)}},
 		env:  []string{"DEPLOY_TYPE=docker-container"},
 		want: false,
-	}, {
-		name: "SearchIndex INDEXED_SEARCH=f",
-		sc:   &Unified{},
-		env:  []string{"DEPLOY_TYPE=docker-container", "INDEXED_SEARCH=f"},
-		want: false,
-	}, {
-		name: "SearchIndex INDEXED_SEARCH=t",
-		sc:   &Unified{},
-		env:  []string{"DEPLOY_TYPE=docker-container", "INDEXED_SEARCH=t"},
-		want: true,
 	}}
 	for _, test := range tests {
 		t.Run(test.name, func(t *testing.T) {

--- a/internal/conf/computed_test.go
+++ b/internal/conf/computed_test.go
@@ -19,10 +19,10 @@ func TestSearchIndexEnabled(t *testing.T) {
 		env  []string
 		want interface{}
 	}{{
-		name: "SearchIndex defaults to false in docker",
+		name: "SearchIndex defaults to true in docker",
 		sc:   &Unified{},
 		env:  []string{"DEPLOY_TYPE=docker-container"},
-		want: false,
+		want: true,
 	}, {
 		name: "SearchIndex defaults to true in k8s",
 		sc:   &Unified{},

--- a/internal/search/zoekt/indexed_search.go
+++ b/internal/search/zoekt/indexed_search.go
@@ -149,7 +149,7 @@ type IndexedSearchRequest interface {
 }
 
 func NewIndexedSearchRequest(ctx context.Context, args *search.TextParameters, typ search.IndexedRequestType, onMissing OnMissingRepoRevs) (IndexedSearchRequest, error) {
-	if args.Mode == search.ZoektGlobalSearch {
+	if args.Zoekt != nil && args.Mode == search.ZoektGlobalSearch {
 		// performance: optimize global searches where Zoekt searches
 		// all shards anyway.
 		return NewIndexedUniverseSearchRequest(ctx, args, typ, args.RepoOptions, args.UserPrivateRepos)

